pkg/reconciler/workload/apply.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workload
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	workloadv1alpha1 "github.com/crossplane/crossplane/apis/workload/v1alpha1"
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	errNotKubeApp            = "object is not a KubernetesApplication"
+	errCreateKubeApp         = "cannot create KubernetesApplication"
+	errGetKubeApp            = "cannot get KubernetesApplication"
+	errDiffController        = "existing KubernetesApplication has a different (or no) controller"
+	errMergeKubeAppTemplates = "cannot merge KubernetesApplicationResourceTemplates"
+	errPatchKubeApp          = "cannot patch existing KubernetesApplication"
+)
+
+// KubeAppApply applies changes to the supplied KubernetesApplication. The
+// KubernetesApplication will be created if it does not exist, or patched if it
+// does. Resource templates are merged instead of replaced if they have the same
+// name and GroupVersionKind.
+func KubeAppApply(ctx context.Context, c client.Client, o runtime.Object, ao ...resource.ApplyOption) error {
```

Comment:
Could this be reimplemented as a custom ApplyOption now that https://github.com/crossplane/crossplane-runtime/pull/129 is merged?

Reply:
Wondering if the `WithApplicator` should be removed from the reconciler, and we always use the `resource.Apply`, but can pass in `AppyOptions`...